httpx/dispatch/connection_pool.py

```diff
@@ -124,7 +126,13 @@ def num_connections(self) -> int:
             self.active_connections.remove(connection)
             self.max_connections.release()
             raise exc
-
+        except NoAvailableStreamIDError:
```

Comment:
I think it'd be nice for us to *not* expose `h2.NoAvailableStreamIDError` to the connection pool,
but rather have an `httpx` exception, such as `ReconnectRequired`.

We're not exposing any other `h2` library implementation details at this layer, and it'd be nice to keep in line with that. (And I can imagine that there could be other circumstances where we'd want the connection to request a reconnect & retry.)